samples/nrf9160/modem_shell/src/link/link_shell.c

```diff
@@ -1043,704 +1266,800 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
+		if (ret < 0) {
+			mosh_error("Cannot read system mode of the modem: %d", ret);
+		} else {
+			enum lte_lc_system_mode sett_sys_mode;
+			enum lte_lc_system_mode_preference sett_lte_pref;
+
+			/* Print also settings stored in mosh side: */
+			link_sett_sysmode_print();
+			sett_sys_mode = link_sett_sysmode_get();
+			sett_lte_pref = link_sett_sysmode_lte_preference_get();
+			if (sett_sys_mode != LTE_LC_SYSTEM_MODE_NONE &&
+				sett_sys_mode != sys_mode_current &&
+				sett_lte_pref != sys_mode_pref_current) {
+				mosh_warn(
+					"note: seems that set link sysmode and a "
+					"counterparts in modem are not in synch");
+				mosh_warn(
+					"but no worries; requested system mode retried "
+					"next time when going to normal mode");
+			}
 		}
+	} else if (sysmode_option != LTE_LC_SYSTEM_MODE_NONE) {
+		link_shell_sysmode_set(sysmode_option, sysmode_lte_pref_option);
 
-		ret = link_setdnsaddr(ip_address ? ip_address : link_sett_dnsaddr_ip_get());
+		/* Save system modem to link settings: */
+		(void)link_sett_sysmode_save(sysmode_option, sysmode_lte_pref_option);
 
-		if (ip_address) {
-			link_sett_save_dnsaddr_ip(ip_address);
-		}
-		break;
-	case LINK_CMD_DISCONNECT:
-		ret = link_shell_pdn_disconnect(pdn_cid);
-		break;
-	default:
-		mosh_error("Internal error. Unknown link command=%d", link_cmd_args.command);
-		ret = -EINVAL;
-		break;
+	} else if (common_option == LINK_COMMON_RESET) {
+		link_shell_sysmode_set(SYS_MODE_PREFERRED, CONFIG_LTE_MODE_PREFERENCE);
+
+		(void)link_sett_sysmode_default_set();
+	} else {
+		link_shell_print_usage(LINK_CMD_SYSMODE);
 	}
-	return ret;
 
-show_usage:
-	/* Reset getopt for another users: */
+	return 0;
+}
+
+static int link_shell_tau(const struct shell *shell, size_t argc, char **argv)
+{
+	enum link_shell_common_options common_option = LINK_COMMON_NONE;
+	int threshold_time = 0;
+
+	int long_index = 0;
+	int opt;
+
 	optreset = 1;
+	optind = 1;
+	while ((opt = getopt_long(argc, argv, short_options, long_options, &long_index)) != -1) {
+		link_shell_getopt_common_option(opt, &common_option);
 
-	link_shell_print_usage(&link_cmd_args);
-	return ret;
+		switch (opt) {
+		case LINK_SHELL_OPT_THRESHOLD_TIME:
+			threshold_time = atoi(optarg);
+			if (threshold_time <= 0) {
+				mosh_error(
+					"Not a valid number for --threshold_time (milliseconds).");
+				return -EINVAL;
+			}
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (common_option == LINK_COMMON_SUBSCRIBE) {
+		link_modem_tau_notifications_subscribe(
+			CONFIG_LTE_LC_TAU_PRE_WARNING_TIME_MS,
+			((threshold_time) ?
+				threshold_time :
+				CONFIG_LTE_LC_TAU_PRE_WARNING_THRESHOLD_MS));
+	} else if (common_option == LINK_COMMON_UNSUBSCRIBE) {
+		link_modem_tau_notifications_unsubscribe();
+	} else {
+		link_shell_print_usage(LINK_CMD_TAU);
+	}
+
+	return 0;
 }
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_link,
+	SHELL_CMD_ARG(
+		coneval, NULL,
+		"Get connection evaluation parameters (no options).",
+		link_shell_coneval, 0, 10),
```

Comment:
All commands have zero mandatory arguments and 10 optional arguments. At least for those commands which do not take arguments, this could be changed so that an error is given if there are arguments present on the command line. Currently possible arguments are ignored and the command is executed anyway.
```suggestion
		link_shell_coneval, 1, 0),
```
This would make the command fail in case there are arguments:
```
mosh:~$ link coneval foo
coneval: wrong parameter count
coneval - Get connection evaluation parameters (no options).
```